tiledb/sm/query/dimension_label/dimension_label_query.h

```diff
@@ -0,0 +1,181 @@
+
+/** Class for dimension label query status exceptions. */
+class DimensionLabelQueryStatusException : public StatusException {
+ public:
+  explicit DimensionLabelQueryStatusException(const std::string& msg)
+      : StatusException("DimensionLabelQuery", msg) {
+  }
+};
+
+/** Dimension label query for writing ordered data. */
+class DimensionLabelQuery : public Query {
+ public:
+  /** Default constructor is not C.41 compliant. */
+  DimensionLabelQuery() = delete;
+
+  /**
+   * Constructor for queries to read or write label data.
+   *
+   * @param storage_manager Storage manager object.
+   * @param name Name of the dimension label.
+   * @param dimension_label Opened dimension label for the query.
+   * @param parent_subarrray Subarray of the parent array.
+   * @param label_buffer Query buffer for the label data.
+   * @param index_buffer Query buffer for the index data. May be empty if no
+   *     index buffer is set.
+   * @param dim_idx Index of the dimension on the parent array this dimension
+   *     label is for.
+   * @param fragment_name Name to use when writing the fragment.
+   */
+  DimensionLabelQuery(
+      StorageManager* storage_manager,
+      stats::Stats* stats,
+      const std::string& name,
+      DimensionLabel* dimension_label,
+      const Subarray& parent_subarray,
+      const QueryBuffer& label_buffer,
+      const QueryBuffer& index_buffer,
+      const uint32_t dim_idx,
+      optional<std::string> fragment_name);
+
+  /**
+   * Constructor for range queries.
+   *
+   * @param storage_manager Storage manager object.
+   * @param dimension_label Opened dimension label for the query.
+   * @param label_ranges Label ranges to read index ranges from.
+   */
+  DimensionLabelQuery(
+      StorageManager* storage_manager,
+      DimensionLabel* dimension_label,
+      const std::vector<Range>& label_ranges);
+
+  /** Disable copy and move. */
+  DISABLE_COPY_AND_COPY_ASSIGN(DimensionLabelQuery);
+  DISABLE_MOVE_AND_MOVE_ASSIGN(DimensionLabelQuery);
+
+  /** Returns ``true`` if the query status is completed. */
+  bool completed() const;
+
+  /** Returns access to the internally stored index data. */
+  inline IndexData* index_data() {
+    return index_data_.get();
+  }
+
+  /** Returns the name of the dimension label. */
+  inline const std::string& name() const {
+    return name_;
+  }
+
+ private:
+  /** The name of the dimension label. */
+  std::string name_;
+
+  /**
+   * Internally managed index data.
+   *
+   * Note: May be null if the index data is set and managed by the user.
+   */
+  tdb_unique_ptr<IndexData> index_data_;
+
+  /**
+   * Initializes the query for reading label data.
+   *
+   * @param dimension_label Opened dimension label for the query.
+   * @param parent_subarrray Subarray of the parent array.
+   * @param label_buffer Query buffer for the label data.
+   * @param dim_idx Index of the dimension on the parent array this dimension
+   */
+  void initialize_read_labels_query(
```

Comment:
This is only to be used during construction. I can add a comment to the effect. The only reason any of these initialize functions exist is to break the constructor into reasonable sized chunks.

Reply:
Ah I see, yes let's add a comment... Although when looking in a normal code editor it will be clearer than looking at a git diff, it will not hurt to add extra doc to the function.